### ISSUE TITLE
define behavior when using type variables

### ISSUE THREAD

**Issue by lukasj** [eclipse-ee4j/jpa-api#142]
Currently when having a base class for entities <tt>BaseEntity</tt> with a type variable it is not defined what the metamodel should report as type for attributes that use a type variable. Let's consider the following example.

```
@MappedSuperclass
public class BaseEntity<T extends Serializable> {
  @EmbeddedId
  private T id;
}

@Embeddable
public class MyEntityId implements Serializable {
  private String someValue;
}

@Entity
public class MyEntity extends BaseEntity<MyEntityId> {
}
```

When asking the metamodel for the id type of <tt>MyEntity</tt> through <tt>EntityType.getIdType()</tt> it is undefined what it should report. Some JPA implementations return the type bound <tt>Serializable</tt>, others return the actual type <tt>MyEntityId</tt> and some even return something completely different <tt>String</tt>.

Analogous to that is that the return of <tt>Attribute.getJavaType()</tt> which is also undefined for the attribute **id**.

I propose to add another method to `javax.persistence.metamodel.Attribute` that allows to resolve the attribute type based on a container class like

```
public interface Attribute<X, Y> {
  default Class<? extends Y> getJavaType(Class<? extends X> containerClass) {
    return getJavaType();
  }
  // other methods }
```

It should be clarified that <tt>getJavaType()</tt> must return the type bound i.e. <tt>Serializable</tt> in this case. The use of the concrete subtype as container class like <tt>idAttribute.getJavaType(MyEntity.class)</tt> should resolve to the correct type <tt>MyEntityId</tt>.
#### Affected Versions
[2.1.1]

**Comment by gavinking** [comment]
This is very related to #562, right, @beikov?

And it would also be solved by the:

```java
Attribute<X, ?> getActualAttribute(String name);
```

method I proposed in [this comment](https://github.com/jakartaee/persistence/issues/562#issuecomment-1862602653), assuming I'm understanding correctly.

**Comment by beikov** [comment]
Correct, it would also solve the underlying problem, though it would require an (actual) `Attribute` to exist, which has no place in the static metamodel at this time. If we go with that API, we should probably also specify that a metamodel generator should emit the actual attribute in the subtype metamodel classes.
The next thing we have to consider is how to deal with type variables in abstract entity types, which is something I started working on recently since it was widely requested: https://hibernate.atlassian.net/browse/HHH-19978

Bottom line, I'll close this in favor of #562

**Comment by gavinking** [comment]
@beikov Let's leave this one open and close #562, since I think this issue description is closer to what we want to do. (And also this issue has priority by quite a few years.)

**Comment by beikov** [comment]
Cross linking the Hibernate ORM request for inclusion in the static metamodel: [HHH-19986](https://hibernate.atlassian.net/browse/HHH-19986)
